cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -711,6 +718,218 @@ class self_comparator {
+class element_hasher {
+ public:
+  __device__ element_hasher(
+    Nullate nulls,
+    uint32_t seed             = DEFAULT_HASH_SEED,
+    hash_value_type null_hash = std::numeric_limits<hash_value_type>::max()) noexcept
+    : _check_nulls(nulls), _seed(seed), _null_hash(null_hash)
+  {
+  }
+
+  template <typename T, CUDF_ENABLE_IF(column_device_view::has_element_accessor<T>())>
+  __device__ hash_value_type operator()(column_device_view const& col,
+                                        size_type row_index) const noexcept
+  {
+    if (_check_nulls && col.is_null(row_index)) { return _null_hash; }
+    return hash_function<T>{_seed}(col.element<T>(row_index));
+  }
+
+  template <typename T, CUDF_ENABLE_IF(not column_device_view::has_element_accessor<T>())>
+  __device__ hash_value_type operator()(column_device_view const& col,
+                                        size_type row_index) const noexcept
+  {
+    CUDF_UNREACHABLE("Unsupported type in hash.");
+  }
+
+  uint32_t _seed;
+  hash_value_type _null_hash;
+  Nullate _check_nulls;
+};
+
+/**
+ * @brief Computes the hash value of a row in the given table.
+ *
+ * @tparam hash_function Hash functor to use for hashing elements.
+ * @tparam Nullate A cudf::nullate type describing the existence of nulls in the table
+ */
+template <template <typename> class hash_function, typename Nullate>
+class device_row_hasher {
+  friend class row_hasher;
+
+ public:
+  device_row_hasher() = delete;
+
+  __device__ auto operator()(size_type row_index) const noexcept
+  {
+    // Hash the first column w/ the seed
+    auto const initial_hash =
+      cudf::detail::hash_combine(hash_value_type{0},
+                                 type_dispatcher<dispatch_storage_type>(
+                                   _table.column(0).type(),
+                                   element_hasher_adapter<hash_function>{_check_nulls, _seed},
+                                   _table.column(0),
+                                   row_index));
+
+    auto it = thrust::make_transform_iterator(_table.begin(), [=](auto const& column) {
+      return cudf::type_dispatcher<dispatch_storage_type>(
+        column.type(), element_hasher_adapter<hash_function>{_check_nulls}, column, row_index);
+    });
+
+    // Hash each element and combine all the hash values together
+    // note that this starts at 1 and not 0 since we already hashed the first column
+    return detail::accumulate(
+      it + 1, it + _table.num_columns(), initial_hash, [](auto hash, auto h) {
+        return cudf::detail::hash_combine(hash, h);
+      });
+  }
+
+ private:
+  /**
+   * @brief Computes the hash value of an element in the given column.
+   *
+   * When the column is non-nested, this is a simple wrapper around the element_hasher.
+   * When the column is nested, this uses the element_hasher to hash the shape and values of the
+   * column.
+   */
+  template <template <typename> class hash_fn>
+  class element_hasher_adapter {
```

Comment:
I could have sworn I already left this as a comment somewhere, but I can't find it now, so I'll risk repeating it.

The `element_hasher_adapter` could be simplified by eliminating the redundancy of repeating the members of `element_hasher` to reconstruct an `element_hasher` each time. Instead, just make the `element_hasher` itself a member of `element_hasher_adapter`:

```
template <typename ElementHasher>
struct element_hasher_adapter{
...
private:
   ElementHasher const hasher_;
}
```

Reply:
This? https://github.com/rapidsai/cudf/pull/10641#discussion_r850565318

You resolved it.

Reply:
Also this? https://github.com/rapidsai/cudf/pull/10641#discussion_r850569846

I resolved the conversation after hardcoding the null hash and pointing the digression towards the other discussion.

Reply:
The suggestion here is different from the two discussions I linked so maybe this is new

Reply:
Yeah, those comments are different. I must have just never submitted the original comment I'm thinking of.